pallets/xcm-asset-manager/src/lib.rs

```diff
@@ -0,0 +1,448 @@
+
+			AssetMetadatas::<T>::insert(&asset_id, &metadata);
+			AssetIdType::<T>::insert(&asset_id, &asset_type);
+			AssetTypeId::<T>::insert(&asset_type, &asset_id);
+
+			Self::deposit_event(Event::<T>::ForeignAssetRegistered { asset_id, asset_type });
+			Self::deposit_event(Event::<T>::ForeignAssetMetadataUpdated { asset_id, metadata });
+
+			// Auto increment for Asset cunter
+			ForeignAssetTracker::<T>::put(asset_id + One::one());
+
+			Ok(())
+		}
+
+		/// Relocate asset id.
+		/// Can only be larger than current assignment.
+		#[pallet::weight(T::WeightInfo::relocate_foreign_asset_id())]
+		pub fn relocate_foreign_asset_id(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			if asset_id > ForeignAssetTracker::<T>::get() {
+				ForeignAssetTracker::<T>::put(asset_id);
+			}
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::update_foreign_asset_metadata())]
+		pub fn update_foreign_asset_metadata(
+			origin: OriginFor<T>,
+			asset_type: T::ForeignAssetType,
+			metadata: AssetMetadata<BalanceOf<T>>,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			let asset_id: T::AssetId =
+				AssetTypeId::<T>::get(&asset_type).ok_or(Error::<T>::AssetTypeDoesNotExist)?;
+			AssetMetadatas::<T>::insert(&asset_id, &metadata);
+
+			Self::deposit_event(Event::<T>::ForeignAssetMetadataUpdated { asset_id, metadata });
+			Ok(())
+		}
+
+		/// Change the amount of units we are charging per execution second
+		/// for a given ForeignAssetType
+		#[pallet::weight(T::WeightInfo::set_asset_units_per_second(*num_assets_weight_hint))]
+		pub fn set_asset_units_per_second(
+			origin: OriginFor<T>,
+			asset_type: T::ForeignAssetType,
+			units_per_second: u128,
+			num_assets_weight_hint: u32,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			let asset_id: T::AssetId =
+				AssetTypeId::<T>::get(&asset_type).ok_or(Error::<T>::AssetTypeDoesNotExist)?;
+
+			// Grab supported assets
+			let mut supported_assets = SupportedFeePaymentAssets::<T>::get();
+
+			ensure!(
+				num_assets_weight_hint >= (supported_assets.len() as u32),
+				Error::<T>::TooLowNumAssetsWeightHint
+			);
+
+			// Only if the asset is not supported we need to push it
+			if let Err(index) = supported_assets.binary_search(&asset_id) {
+				supported_assets.insert(index, asset_id.clone());
+				SupportedFeePaymentAssets::<T>::put(supported_assets);
+			}
+
+			AssetIdUnitsPerSecond::<T>::insert(&asset_id, &units_per_second);
+
+			Self::deposit_event(Event::UnitsPerSecondChanged { asset_id, units_per_second });
+			Ok(())
+		}
+
+		/// Add the xcm type mapping for a existing assetId, other assetType still exists if any.
+		/// TODO: Weight is no longer related to num_assets_weight_hint; changes here needed
+		#[pallet::weight(T::WeightInfo::add_asset_type())]
+		pub fn add_asset_type(
```

Comment:
It's more like `update_assert_type` to me, the parameter doesn't seem to be consistent either: we are using `asset_id` while other extrinsics use `asset_type` to find the element.

Reply:
a AssetId only mappings to a special AssetType.
Many AssetType can mappings to one same AssetId.

Reply:
Actually, I do hope to have 1-to-1 mapping plus N-to-1 mapping.
In XCM system, it is necessary to keep compatiable of different multilocation standard.

We can observe many standard so far. ( e.g. <ParachainId(X), Here>, <ParachainId(X), PalletInstance(pallet_balances index)>, <ParachainId(X), GeneralKey(***)>, <ParachainId(X)>)

https://moonriver.subscan.io/event?address=&module=assetmanager&event=assetregistered&startDate=&endDate=&startBlock=&endBlock=&timeType=date&version=1201

The thing I am thinking is, In most circumstances, other parachain change style, we adapt immediately and forget about old style, it is fine. But suppose Parachain A sending a old style of XCM of token on Parachain B to Parachain C (since Parachain A does not update Parachain B's style immediately), Parachain C (that is us) will not be able to recognize it.

And Multilocation compatibility is a common request I think. So N-to-1 mapping I hope it will be worth keeping.

But yes, I think I can work around the merge of add/remvoe_asset_type into one extrinsic. It sounds good.